File: java/binarySearch/src/main/java/BinarySearch.java
/**
 * 查找第一个值等于给定值的元素
 *
 * 查找最后一个值等于给定值的元素
 *
 * 查找第一个大于等于给定值的元素
 *
 * 查找最后一个小于等于给定值的元素
 */
public class BinarySearch {

    /**
     * 查找第一个值等于给定值的元素
     * @param a 有序数组
     * @param item 查找元素
     * @param n 数据规模
     * @return 给定元素下标，没找到返回-1
     */
    public int findFirstEqual(int[] a,int item, int n){
        int low = 0;
        int high = n-1;
        if(low==high){
            return item == a[low] ? low : -1;
        }
        while(low<=high){
            int mid = low + ((high - low) >> 2);
            if(a[mid]<item){
                low = mid + 1;
            } else if(a[mid] > item){
                high = mid - 1;
            }else{
                if(mid == 0 || a[mid - 1]!=item) return mid;
                high = mid - 1;
            }
        }
        return -1;

    }

    /**
     * 查找最后一个值等于给定值的元素
     * @param a 有序数组
     * @param item 查找元素
     * @param n 数据规模
     * @return 给定元素下标，没找到返回-1
     */
    public int findLastEqual(int[] a, int item, int n){
        int low = 0;
        int high = n-1;
        if(low == high){
            return item == a[low]? low: -1;
        }
        while(low<=high){
            int mid = low + ((high-low)>>2);
            if(a[mid]<item){
                low = mid + 1;
            } else if(a[mid] > item){
                high = mid - 1;
            }else{
                if(mid == n-1 || a[mid+1]!=item) return mid;
                low = mid + 1;
            }
        }
        return -1;

    }

    /**
     * 查找第一个值大于等于给定值的元素
     * @param a 有序数组
     * @param item 查找元素
     * @param n 数据规模
     * @return 给定元素下标，没找到返回-1
     */
    public int findFirstGe(int[] a, int item, int n){
        int low = 0;
        int high = n -1;
        if(low==high) return a[low]==item?low:-1;
        while(low<=high){
            int mid = low + ((high - low )>>2);
            if(a[mid]<item){
                low = mid + 1;
            }else{
                if(mid == 0 || a[mid-1]<item) return mid;
                high = mid -1;
            }
        }
        return -1;
    }
    /**
     * 查找最后一个值小于等于给定值的元素
     * @param a 有序数组
     * @param item 查找元素
     * @param n 数据规模
     * @return 给定元素下标，没找到返回-1
     */
    public int findLastLe(int[] a, int item, int n){
        int low = 0;
        int high = n - 1;
        if(low == high) return a[low] == item ? low : -1;
        while(low<=high){
            int mid = low + ((high - low) >> 2);
            if(a[mid]<=item){
                if(mid == n-1 || a[mid+1] > item) return mid;
                low = mid + 1;
            }else{
                high = mid -1;
            }
        }
        return -1;
    }
}

File: Python/binarySearch/BinarySearch.py
# -*- coding:utf-8 -*-

import random


class BinarySearch:
    def __init__(self, arr):
        self.__arr__ = sorted(arr)

    def binary_search(self, item):
        arr = self.__arr__
        high = len(arr) - 1
        low = 0
        while low <= high:
            mid = low + ((high-low) >> 2)
            if item < arr[mid]:
                high = mid - 1
            elif item > arr[mid]:
                low = mid + 1
            else:
                return mid
        return -1

    def first_equal(self, item):
        arr = self.__arr__
        high = len(arr) - 1
        low = 0
        while low <= high:
            mid = low + ((high - low) >> 2)
            if item < arr[mid]:
                high = mid - 1
            elif item > arr[mid]:
                low = mid + 1
            else:
                if mid == 0 or arr[mid - 1] != item:
                    return mid
                high = mid - 1
        return -1

    def last_eq(self, item):
        arr = self.__arr__
        high = len(arr) - 1
        low = 0
        while low <= high:
            mid = low + ((high - low) >> 2)
            if item < arr[mid]:
                high = mid - 1
            elif item > arr[mid]:
                low = mid + 1
            else:
                if mid == len(arr) - 1 or arr[mid + 1] != item:
                    return mid
                low = mid + 1
        return -1

    def last_lt(self, item):
        arr = self.__arr__
        high = len(arr) - 1
        low = 0
        while low <= high:
            mid = low + ((high - low) >> 2)
            if item <= arr[mid]:

                high = mid - 1
            else:
                if mid == len(arr) - 1 or arr[mid + 1] > item:
                    return mid
                low = mid + 1
        return -1

    def first_gt(self, item):
        arr = self.__arr__
        high = len(arr) - 1
        low = 0
        while low <= high:
            mid = low + ((high - low) >> 2)
            if item <= arr[mid]:
                if mid == 0 or arr[mid - 1] < item:
                    return mid
                high = mid - 1
            else:

                low = mid + 1
        return -1

    def display(self):
        print(self.__arr__)


if __name__ == '__main__':
    a = [1, 2, 2, 4, 4, 4, 5]
    print(a)
    bs = BinarySearch(a)
    bs.display()
    print(bs.binary_search(a[2]))
    print(bs.first_equal(2))
    print(bs.last_eq(4))
    print(bs.last_lt(3))
    print(bs.first_gt(3))